Client interactive suggest (extract info from CREATE queries)

azat opened the pull request:
Changelog category (leave one):
- Not for changelog (changelog entry is not required)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Client interactive suggest (extract info from CREATE queries)

Detailed description / Documentation draft:
This will parse CREATE queries and add the following things to
completion list for clickhouse-client/clickhouse-local:
- table
- database
- columns

Supersedes: #32370

azat committed:
Client interactive suggest (extract info from CREATE queries)

This will parse CREATE queries and add the following things to
completion list for clickhouse-client/clickhouse-local:
- table
- database
- columns

azat committed:
Fix 02160_client_autocomplete_parse_query

Sad (since I wanted to make the test a little bit cleaner/easier to
understand), but you cannot use send_user in tests, since expect
send_user works only if stdin is a tty, so the following will fail (so
as in CI):

    $ ./tests/clickhouse-test --database default 02160  < /dev/null
    Using queries from '/src/ch/clickhouse-upstream/tests/queries' directory
    Using clickhouse-client as client program (expecting split build)
    Connecting to ClickHouse server... OK

    Running 1 stateless tests (MainProcess).

    02160_client_autocomplete_parse_query:                                  [ FAIL ] - return code: 1
    send: spawn id exp0 not open
        while executing
    "send_user "Completion loaded\n""
        (file "/src/ch/clickhouse-upstream/tests/queries/0_stateless/02160_client_autocomplete_parse_query.expect" line 40)
    , result:

    stdout:

    Database: default

azat committed:
Fix 02160_client_autocomplete_parse_query by waiting for the client to be spawned

expect can be terminated with eof, if the process has not been started
yet, add a while loop to wait until it will be started.

azat committed:
Fix 02160_client_autocomplete_parse_query in a more better way

The problem was that "eof" handler had been reseted.

azat commented:
>[Stress test (address, actions) — Fatal message in clickhouse-server.log](https://s3.amazonaws.com/clickhouse-test-reports/33201/4f04d6af61eabf4899eb8188150dc862aaab80fc/stress_test__address__actions_.html)

- #33386

azat commented:
@vitlibar friendly ping

vitlibar commented on base/base/LineReader.cpp:
```diff
@@ -65,6 +65,27 @@ std::optional<LineReader::Suggest::WordsRange> LineReader::Suggest::getCompletio
+            continue;
+
+        words.push_back(new_word);
+        words_no_case.push_back(new_word);
```
For me it doesn't look reliable. You're going to modify the vector while other function of the `Suggest` structure returns iterators inside this vector. It doesn't look very safe. Also you don't call `updateSuggest()` until the main completion list has been loaded from the server and it doesn't sound very reasonable. What I suggest is to make `LineReader::Suggest` a well-rounded class with a mutex so it could add words from multiple threads safely:

```
class Suggest
{
public:
    using Words = std::vector<std::string>;

    /// Called by Suggest::load() and also by ClientBase::updateSuggest().
    void addWords(Words && new_words);
    
    /// We don't give away any internal iterators
    void getCompletions(Replxx::completions_t & completions, const String & prefix, size_t prefix_length) const;

private:
    Words words;
    Words words_no_case;
    std::mutex mutex;
    /// We don't need std::atomic<bool> ready{false} anymore!
};
```

azat replied:
> For me it doesn't look reliable. You're going to modify the vector while other function of the Suggest structure returns iterators inside this vector. It doesn't look very safe.

getCompletions() returns iterator, but it is converted to `Replxx::completions_t` instantly

https://github.com/ClickHouse/ClickHouse/blob/1e8e21570ccb3b557588d25e191c969efcf2efd0/base/base/ReplxxLineReader.cpp#L182-L183

I haven't touched this to reduce the patch.
Will convert getCompletion to return vector over iterator range. 

And addWords should not be run in parallel with getCompletions, so this should be safe.

> Also you don't call updateSuggest() until the main completion list has been loaded from the server and it doesn't sound very reasonable.

Yes, otherwise Suggest should store newly added words separately, but I did not want to introduce extra complexity for this simple thing, since usually when user type something in the client suggestion is already loaded.

> What I suggest is to make LineReader::Suggest a well-rounded class with a mutex so it could add words from multiple threads safely:

I don't think that it requires lock, since like I said, addWords should not be called in parallel with getCompletions, but if you insist, I can go this way.

vitlibar replied:
Yes, I believe your code works, I mean it just doesn't look right. I mean using the variable `std::atomic<bool> ready` to synchonize stuff didn't look very nice even before your patch, and now you're adding complexity and it's going to make the things a bit less understandable. That's why I suggest to make one step further and make synchronization more obvious and easy to see.

azat replied:
Ok, I've replaced that atomic with mutex.

vitlibar commented on tests/queries/0_stateless/02160_client_autocomplete_parse_query.expect:
```diff
@@ -0,0 +1,81 @@
+    }
+}
+set timeout 60
+send -- ""
```
What is the character?

azat replied:
Ctrl-C

vitlibar replied:
Can we escape it here somehow, or at least write a comment about that?

azat replied:
Sure, done.

azat committed:
Use hex char numbers for Ctrl-C/Ctrl-V in 02160_client_autocomplete_parse_query

Since special symbols for Ctrl-C/Ctrl-V are not rendered in github UI,
yes vim renders it, but this does not makes review easier.

azat committed:
Rewrite suggestions to make it more cleaner

- use plain mutex over atomic
- use addWords() for initial suggestion filling
- return vector<> from getCompletions() over iterator, to avoid possible
  questions (even though it was safe, since addWords() could not be
  called in parallel with getCompletions() before)

vitlibar commented on base/base/LineReader.cpp:
```diff
@@ -36,33 +36,67 @@ bool hasInputData()
+        {
+            return std::tolower(char1) < std::tolower(char2);
+        });
+    });
```
This way you're going to sort all the words again even if you've just added a single word to the list. We could instead put new words to the end of `words` without sorting, then sort only new words, after that use `std::inplace_merge` to get the full list sorted and finally call `std::unique()` to remove duplicates.

azat replied:
I thought about `std::inplace_merge` initially, but decided that it does not worth it, since CREATE queries are not common.

But I like that after `std::inplace_merge`, `addWords()` can be used everywhere, so let's go this way.

vitlibar commented on src/Client/Suggest.cpp:
```diff
@@ -29,19 +29,21 @@ namespace ErrorCodes
+        "PROFILE",      "QUOTA",    "POLICY", "ROW",       "GRANT",    "REVOKE",      "OPTION",   "ADMIN",       "EXCEPT",  "REPLACE",
+        "IDENTIFIED",   "HOST",     "NAME",   "READONLY",  "WRITABLE", "PERMISSIVE",  "FOR",      "RESTRICTIVE", "RANDOMIZED",
+        "INTERVAL",     "LIMITS",   "ONLY",   "TRACKING",  "IP",       "REGEXP",      "ILIKE",
+    };
```
Why didn't you call `addWords()` for that list?

azat replied:
Because before addWords has optimization only for initial load, and this will make initial load a little bit more expensive, not a problem with the current implementation anyway.

vitlibar commented on base/base/ReplxxLineReader.cpp:
```diff
@@ -179,9 +179,8 @@ ReplxxLineReader::ReplxxLineReader(
-        if (auto range = suggest.getCompletions(context, context_size))
-            return Replxx::completions_t(range->first, range->second);
-        return Replxx::completions_t();
+        auto words = suggest.getCompletions(context, context_size);
```
Probably it's better to return `Replxx::completions_t` directly from `getCompletions()`. This is because `Replxx::completions_t` is a vector of `Replxx::Completion`, and `Replxx::Completion` keeps a string inside. So doing this way you're going to copy all the completion twice - the first time while you're collecting words inside `getCompletions()` and the second time while you're constructing `Replxx::completions_t` here.

azat replied:
>Probably it's better to return Replxx::completions_t directly from getCompletions(). 

This will break abstractions a little, but I'm fine with this.

>This is because Replxx::completions_t is a vector of Replxx::Completion, and Replxx::Completion keeps a string inside

Right.

vitlibar commented on src/Client/ClientBase.cpp:
```diff
@@ -47,6 +47,10 @@
+#include <Parsers/ASTColumnDeclaration.h>
+#include <Parsers/ASTIndexDeclaration.h>
+#include <Parsers/ASTConstraintDeclaration.h>
+#include <Parsers/ASTProjectionDeclaration.h>
```
Why do we need anything except `ASTColumnDeclaration.h` here?

azat replied:
Nice catch, that was leftovers from the first version.

vitlibar commented on src/Client/Suggest.cpp:
```diff
@@ -191,7 +183,7 @@ void Suggest::fillWordsFromBlock(const Block & block)
     size_t rows = block.rows();
     for (size_t i = 0; i < rows; ++i)
-        words.emplace_back(column.getDataAt(i).toString());
+        new_words.emplace_back(column.getDataAt(i).toString());
```
Can we just call `addWords()` here and get rid of necessity of having `new_words` as a persistent variable.

azat replied:
Decided to put make new_words local variable, to process words by batches at least.

azat committed:
Add inplace merge for new completion words

This will allow:
- use addWords() everywhere
- remove that optimization for empty words case (initial load)
- and now we remove duplicates according to comparator

v2: replace parameter pack in addWords() with explicit Compare for both
cases (just pass default)

azat committed:
Remove extra copying for completion words

Return Replxx::completions_t over Words array, that way we avoids one
copying of std::string (although it breaks incapsulation).

vitlibar commented on base/base/LineReader.cpp:
```diff
@@ -34,35 +35,88 @@ bool hasInputData()
+
+void LineReader::Suggest::addWords(Words && new_words)
+{
+    Words new_words_no_case;
+    {
+        new_words_no_case.reserve(new_words.size());
+        new_words_no_case.insert(new_words_no_case.end(), new_words.begin(), new_words.end());
+    }
```
```suggestion
    Words new_words_no_case = new_words;
```

azat commented:
>[Fast test (actions) — fail: 1, passed: 3231, skipped: 545](https://s3.amazonaws.com/clickhouse-test-reports/33201/0013a2fdee26fc11a34ebffaf10f69d7cfff9b3a/fast_test__actions_.html)

- `00821_distributed_storage_with_join_on` - #33976

azat commented:
>[Stateless tests (debug, actions) [2/3] — fail: 1, passed: 1255, skipped: 18](https://s3.amazonaws.com/clickhouse-test-reports/33201/4161c0d9cab496fc8932f8ac0179a9c519120fe3/stateless_tests__debug__actions__[2/3].html)

```
02049_clickhouse_local_merge_tree | FAIL | 2.39
-- | -- | --
2022-01-25 19:33:32 send: spawn id exp3 not open 2022-01-25 19:33:32     while executing 2022-01-25 19:33:32 "send -- "insert into tt values ('2020-01-01', 2, 2), ('2020-01-01', 1, 1)\r"" 2022-01-25 19:33:32     (file "/usr/share/clickhouse-test/queries/0_stateless/02049_clickhouse_local_merge_tree.expect" line 33) 2022-01-25 19:33:32 , result: 2022-01-25 19:33:32  2022-01-25 19:33:32  2022-01-25 19:33:32  2022-01-25 19:33:32 stdout: 2022-01-25 19:33:32  2022-01-25 19:33:32  2022-01-25 19:33:32 Database: test_tn3h3c
```

Interesting, it fails in 2.4 seconds, so this means that it is not timeout, but I cannot reproduce the problem (using local build and binary from CI too).

Let's try again (I've tried to add some debugging in case of failure, but this is not that easy for `expect`).

azat commented:
>[PullRequestCI / SplitBuildSmokeTest (pull_request) Failing after 31s — SplitBuildSmokeTest](https://github.com/ClickHouse/ClickHouse/runs/4966555732?check_suite_focus=true)

```
github.GithubException.GithubException: 502 {"message": "Server Error"}
```

>[PullRequestCI / BuilderSpecialReport (pull_request) Failing after 24s — BuilderSpecialReport](https://github.com/ClickHouse/ClickHouse/runs/4966555580?check_suite_focus=true)

```
github.GithubException.GithubException: 502 {"message": "Server Error"}
```

azat commented:
>[PullRequestCI / FunctionalStatelessTestFlakyCheck (pull_request) — FunctionalStatelessTestFlakyCheck](https://github.com/ClickHouse/ClickHouse/runs/4970016160?check_suite_focus=true)

```
This check failed
```

So even 5 retries did not helped, but this failed check should not be a problem, besides it passed somehow:

>[Stateless tests flaky check (address, actions) — fail: 0, passed: 100](https://s3.amazonaws.com/clickhouse-test-reports/33201/1acd644040b726479f23e527e7059d19e4fa3817/stateless_tests_flaky_check__address__actions_.html)
